### PR TITLE
fix: Added types for runtime config.

### LINKED ISSUE #14392: We need to provide types for resolved config object
Cypress v6.2.0

If you look at all properties in the object returned by `Cypress.config()` call, there are several run-time properties beyond the current typescript interface. For example, the `isInteractive` property is missing from the typescript interface, but present at run-time.

https://docs.cypress.io/guides/references/configuration.html#isInteractive

We should create an extended ConfigAtRuntime typescript interface with these additional properties

### PR DESCRIPTION
**Pull request by sainthkh**
- Closes #14392

### User facing changelog

Added types for config properties appended at runtime.

### Additional details
- Why was this change necessary? => Some runtime config properties are not defined in the type file. 
- What is affected by this change? => N/A

#### Any implementation details to explain? 

I got the list of runtime config properties with the code below:

```js
it('t', () => {
  console.log(Cypress.config())
})
```

And compared the list of options in `server/lib/config_options.ts`.

### How has the user experience changed?

N/A

### Questions

* I intentionally left some explanations for some properties as `???` because it's hard to guess or track the meaning of them.
* I intentionally grouped and left no explanation for internal or unlisted properties. 

### PR Tasks

I intentionally didn't add tests because config types are not tested in the `cli` tests.

### CODE REVIEW

File: cli/types/cypress.d.ts [hunk 1]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+     *
+     * @see https://nodejs.org/api/os.html#os_os_arch
+     */
+    arch: string
+    /**
+     * The list of hosts to be blocked
+     */
+    blockHosts: null | string | string[]
+    /**
+     * The browser Cypress is runnong on.
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2627]:
```suggestion
     * The browser Cypress is running on.
```

File: cli/types/cypress.d.ts [hunk 2]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+     */
+    browser: Browser
+    /**
+     * The browsers Cypress can use.
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2631]:
```suggestion
     * Available browsers found on your system.
```

File: cli/types/cypress.d.ts [hunk 3]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+     */
+    browsers: Browser[]
+    /**
+     * Path to folder containing component test files.
+     */
+    componentFolder: string
+    /**
+     * Enables component testing.
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2639]:
```suggestion
     * Whether component testing is enabled.
```

File: cli/types/cypress.d.ts [hunk 4]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+     */
+    experimentalComponentTesting: boolean
+    /**
+     * Enables polyfill fetch.
+     */
+    experimentalFetchPolyfill: boolean
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2645]: This is deprecated, so probably wouldn't bother documenting this.

```suggestion
```

File: cli/types/cypress.d.ts [hunk 5]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+    /**
+     * ???
+     */
+    hosts: null | string[]
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2649]: I think this is also mostly used internally? We set it in our Cypress config, but I"m not sure why someone else would need this.

File: cli/types/cypress.d.ts [hunk 6]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+    /**
+     * With this option enabled - Cypress will search through the response streams
+     * coming from your server on .html and .js files
+     * and replace code that matches some patterns.
+     *
+     * @see https://docs.cypress.io/guides/references/configuration.html#modifyObstructiveCode
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2655]: on link
```suggestion
     * @see https://on.cypress.io/configuration#modifyObstructiveCode
```

File: cli/types/cypress.d.ts [hunk 7]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+     */
+    experimentalFetchPolyfill: boolean
+    /**
+     * ???
+     */
+    hosts: null | string[]
+    /**
+     * With this option enabled - Cypress will search through the response streams
+     * coming from your server on .html and .js files
+     * and replace code that matches some patterns.
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2653]:
```suggestion
     * Whether Cypress will search for and replace 
     * obstructive JS code in .js or .html files.
```

File: cli/types/cypress.d.ts [hunk 8]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+     *
+     * @see https://docs.cypress.io/guides/references/configuration.html#modifyObstructiveCode
+     */
+    modifyObstructiveCode: boolean
+    /**
+     * ???
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2659]:
```suggestion
     * A unique ID for the project used for recording
```

File: cli/types/cypress.d.ts [hunk 9]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+     */
+    projectId: null | string
+    /**
+     * The path to the support folder
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2663]:
```suggestion
     * Path to the support folder.
```

File: cli/types/cypress.d.ts [hunk 10]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+     */
+    supportFolder: string
+    /**
+     * Glob pattern for the test files.
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2667]:
```suggestion
     * Glob pattern to determine what test files to load.
```

File: cli/types/cypress.d.ts [hunk 11]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+     */
+    testFiles: string
+    /**
+     * ???
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2671]:
```suggestion
     * The user agent the browser sends in all request headers.
```

File: cli/types/cypress.d.ts [hunk 12]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+
+    // Internal or Unlisted at server/lib/config_options
+    autoOpen: boolean
+    browserUrl: string
+    clientRoute: string
+    configFile: string
+    cypressEnv: string
+    integrationExampleName: string
+    integrationExamplePath: string
+    isInteractive: boolean
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2687]: This one is ok to document, we use it in a lot of examples.
```suggestion
    /**
     * Whether Cypress was launched via 'cypress open' (interactive mode)
     */
    isInteractive: boolean
```

File: cli/types/cypress.d.ts [hunk 13]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+    isNewProject: boolean
+    isTextTerminal: boolean
+    morgan: boolean
+    namespace: string
+    parentTestsFolder: string
+    parentTestsFolderDisplay: string
+    platform: 'linux' | 'darwin' | 'win32'
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2694]:
```suggestion
     /**
     * The platform Cypress is running on.
     */
    platform: 'linux' | 'darwin' | 'win32'
```

File: cli/types/cypress.d.ts [hunk 14]
```diff
@@ -2609,6 +2609,109 @@ declare namespace Cypress {
...
+    projectName: string
+    projectRoot: string
+    proxyUrl: string
+    report: boolean
+    reporterRoute: string
+    reporterUrl: string
+    socketId: null | string
+    socketIoCookie: string
+    socketIoRoute: string
+    spec: {
```
Comment (jennifer-shehane) [inline cli/types/cypress.d.ts:2704]: I could see this also being useful to document.

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]

**Comment by sainthkh** [comment]
@jennifer-shehane Thank you for the feedback. I applied them.

File: cli/types/cypress.d.ts (hunk 5 above)
Comment (jennifer-shehane) [reply cli/types/cypress.d.ts:2645]: well, TIL. 

```suggestion
    /**
     * Hosts mappings to IP addresses.
     */
    hosts: null | string[]
```

**Review by jennifer-shehane — APPROVED** [review]

**Pull request merged**